src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs

```diff
@@ -760,30 +911,63 @@ async Task Core(MultiConnectSocketAsyncEventArgs internalArgs, Task<IPAddress[]>
-                        attemptSocket.ReplaceHandleIfNecessaryAfterFailedConnect();
+
+                        attemptSocket2?.ReplaceHandleIfNecessaryAfterFailedConnect();
 
                         // Reconfigure the internal event args for the new address.
-                        if (internalArgs.RemoteEndPoint is IPEndPoint existing)
-                        {
-                            existing.Address = address;
-                            Debug.Assert(existing.Port == port);
-                        }
-                        else
+                        if (address != null)
                         {
-                            internalArgs.RemoteEndPoint = new IPEndPoint(address, port);
+                            Debug.Assert(attemptSocket != null);
+                            attemptSocket.ReplaceHandleIfNecessaryAfterFailedConnect();
+                            if (internalArgs.RemoteEndPoint is IPEndPoint existing)
+                            {
+                                existing.Address = address;
+                                Debug.Assert(existing.Port == port);
+                            }
+                            else
+                            {
+                                internalArgs.RemoteEndPoint = new IPEndPoint(address, port);
+                            }
                         }
 
                         // Issue the connect.  If it pends, wait for it to complete.
-                        if (attemptSocket.ConnectAsync(internalArgs))
+                        bool? result = attemptSocket?.ConnectAsync(internalArgs);
+                        if (result != false)
                         {
+                            if (attemptSocket2 != null)
+                            {
+                                Debug.Assert(internalArgs.SecondarySaea != null);
+                                attemptSocket2.ConnectAsync(internalArgs.SecondarySaea);
+                            }
+
                             using (cancellationToken.UnsafeRegister(s => Socket.CancelConnectAsync((SocketAsyncEventArgs)s!), internalArgs))
                             {
-                                await new ValueTask(internalArgs, internalArgs.Version).ConfigureAwait(false);
+                                //var tasks = new List<ValueTask> { new ValueTask(internalArgs, internalArgs.Version) };
+                                if (internalArgs.SecondarySaea != null)
+                                {
+                                    //var tasks = new List<ValueTask> { new ValueTask(internalArgs, internalArgs.Version) };
+                                    var t1 = new ValueTask(internalArgs, internalArgs.Version).AsTask();
+                                    var t2 = new ValueTask(internalArgs.SecondarySaea, internalArgs.SecondarySaea.Version).AsTask();
+                                    await Task.WhenAny(t1, t2).ConfigureAwait(false);
```

Comment:
Looks like `ConnectAlgorithm.Parallel` brings allocation cost of 3+ `Task` instances. Are we OK with that, assuming we eventually want `SocketsHttpHandler` to utilize it?

Reply:
This was originally simpler .. but had problems in some corner cases. Maybe @stephentoub  has some suggestion how to handle it more efficiently. 

But I don't think we really want to use it in `SocketsHttpHandler` - at least not by default. The parallel connect is way more expensive general - allocates more sockets and that includes OS handles. So what you trade is extra resource cost agains expediency e.g. time to connect (and in some cases reliability as IPv6 can be broken in certain setups)

Reply:
> at least not by default

There is a long discussion on #26177 and triage comment [this triage comment](https://github.com/dotnet/runtime/issues/26177#issuecomment-540141586) implies to me that we actually wanted to consider that. It could help the Connect to "break out" on systems where IPv6 is broken, see #117082 for example. At least it should be made easy (#117548) and as much efficient as possible.

This PR doesn't need to be blocked on perf concerns, but it would be nice to clarify the long term strategy.